### PR TITLE
ref(dashboards): Generic widget component for event queries

### PR DESCRIPTION
**Pull request by narsaynorath**
This adds a Generic Widget Component to conduct widget queries. I tried to add it as a functional component but there were difficulties handling the async behaviour in tests and maintaining the `componentDidUpdate` behaviour (the `componentDidUpdate` has some `setState` logic which I couldn't reproduce easily in a functional component.

I wanted to put this up even as a class component because it'll let us move forward with the other dataset query components.

### CODE REVIEW

File: static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+  loading: boolean;
+  errorMessage?: GenericWidgetQueriesChildrenProps['errorMessage'];
+  pageLinks?: GenericWidgetQueriesChildrenProps['pageLinks'];
+  queryFetchID?: symbol;
+  rawResults?: SeriesResponse[];
+  tableResults?: GenericWidgetQueriesChildrenProps['tableResults'];
+  timeseriesResults?: GenericWidgetQueriesChildrenProps['timeseriesResults'];
+};
+
+class GenericWidgetQueries<SeriesResponse, TableResponse> extends Component<
```
Comment (evanpurkhiser) [inline static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx:75]: In the future would it make sense for this to become just a hook instead?

Comment (narsaynorath) [reply static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx:75]: Hmm, good point. I guess it's just a querying component and right now I don't see why that logic can't be contained in a hook and the right properties returned. It wouldn't be until all of its dependencies are converted to functional components that we'd be able to do that, right? We still have to convert a few and see how this works in other use cases.

File: static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+          api,
+          query,
+          organization,
+          selection,
+          requestLimit,
+          cursor,
+          getReferrer(widget.displayType)
+        );
+      })
+    );
```
Comment (shruthilayaj) [inline static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx:201]: should we try / throw errors here if the `config.getTableRequest` or `config.getWorldMapRequest` is expected but is undefined?

**Review by shruthilayaj — APPROVED** [review]

File: static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx (hunk 1 above)
Comment (evanpurkhiser) [reply static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx:75]: yeah, you're correct. BUT, you could take all of this and put it in a hook, and make a single render props style component that wraps the hook :)

File: static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx (hunk 2 above)
Comment (narsaynorath) [reply static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx]: I added this:
```js
        if (!requestCreator) {
          throw new Error(
            t('This display type is not supported by the selected dataset.')
          );
        }
```

and the try/catch of the parent `fetchData` call pulls out the error message

File: static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+  config: DatasetConfig<SeriesResponse, TableResponse>;
+  organization: Organization;
+  selection: PageFilters;
+  widget: Widget;
+  cursor?: string;
+  limit?: number;
+  onDataFetched?: ({
+    tableResults,
+    timeseriesResults,
+    issuesResults,
```
[the remark this answers is not in the document]
Comment (narsaynorath) [reply static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx:53]: Yeah, good point. I don't really need them here but I removed them all in the other PR at #36103

**Pull request merged**